src/providers/arcgisrest/qgsafsshareddata.cpp

```diff
@@ -159,7 +241,346 @@ QgsFeatureIds QgsAfsSharedData::getFeatureIdsInExtent( const QgsRectangle &exten
+{
+  error.clear();
+  QUrl queryUrl( mDataSource.param( QStringLiteral( "url" ) ) + "/updateFeatures" );
+
+  QgsArcGisRestContext context;
+  context.setObjectIdFieldName( mObjectIdFieldName );
+
+  QgsArcGisRestUtils::FeatureToJsonFlags flags;
+  if ( includeGeometries )
+    flags |= QgsArcGisRestUtils::FeatureToJsonFlag::IncludeGeometry;
+  if ( includeAttributes )
+    flags |= QgsArcGisRestUtils::FeatureToJsonFlag::IncludeNonObjectIdAttributes;
+
+  QVariantList featuresJson;
+  featuresJson.reserve( features.size() );
+  for ( const QgsFeature &feature : features )
+  {
+    featuresJson.append( QgsArcGisRestUtils::featureToJson( feature, context, QgsCoordinateReferenceSystem(), flags ) );
+  }
+
+  const QString json = QString::fromStdString( QgsJsonUtils::jsonFromVariant( featuresJson ).dump( 2 ) );
+
+  QByteArray payload;
+  payload.append( QStringLiteral( "f=json&features=%1" ).arg( json ).toUtf8() );
+
+  bool ok = false;
+  const QVariantMap results = postData( queryUrl, payload, feedback, ok, error );
+  if ( !ok )
+  {
+    return false;
+  }
+
+  const QVariantList addResults = results.value( QStringLiteral( "updateResults" ) ).toList();
+  for ( const QVariant &result : addResults )
+  {
+    const QVariantMap resultMap = result.toMap();
+    if ( !resultMap.value( QStringLiteral( "success" ) ).toBool() )
+    {
+      error = resultMap.value( QStringLiteral( "error" ) ).toMap().value( QStringLiteral( "description" ) ).toString();
+      return false;
+    }
+  }
+
+  // all good. Now we remove the cached versions of features so that they'll get re-fetched from the service
+  QgsReadWriteLocker locker( mReadWriteLock, QgsReadWriteLocker::Write );
+  for ( const QgsFeature &feature : features )
+  {
+    mCache.remove( feature.id() );
+  }
+  return true;
+}
+
+bool QgsAfsSharedData::addFields( const QString &adminUrl, const QList<QgsField> &attributes, QString &error, QgsFeedback *feedback )
+{
+  error.clear();
+  QUrl queryUrl( adminUrl + "/addToDefinition" );
+
+  QVariantList fieldsJson;
+  fieldsJson.reserve( attributes.size() );
+  for ( const QgsField &field : attributes )
+  {
+    fieldsJson.append( QgsArcGisRestUtils::fieldDefinitionToJson( field ) );
+  }
+
+  const QVariantMap definition {{ QStringLiteral( "fields" ), fieldsJson }};
+
+  const QString json = QString::fromStdString( QgsJsonUtils::jsonFromVariant( definition ).dump( 2 ) );
+
+  QByteArray payload;
+  payload.append( QStringLiteral( "f=json&addToDefinition=%1" ).arg( json ).toUtf8() );
+
+  bool ok = false;
+  const QVariantMap results = postData( queryUrl, payload, feedback, ok, error );
+  if ( !ok )
+  {
+    return false;
+  }
+
+  if ( !results.value( QStringLiteral( "success" ) ).toBool() )
+  {
+    error = results.value( QStringLiteral( "error" ) ).toMap().value( QStringLiteral( "message" ) ).toString();
+    return false;
+  }
+
+  // all good. Now we remove the cached versions of features so that they'll get re-fetched from the service
```

Comment:
If you don't mind, you go search and replace `// all good. Now...` with `// All good, now...` :)